### PR TITLE
Xcode workspace and project building.

### PR DESCRIPTION
**Pull request by nicklanasa**
Clickable error links in compilation-mode.
Added method to clean workspaces.
Added method to run pod install.

### CODE REVIEW

File: swift-mode.el [hunk 1]
```diff
@@ -693,6 +693,115 @@ We try to constraint those lookups by reasonable number of lines.")
                "warning: " (message) line-end))
      :modes swift-mode))
 
+;; BUILDING
+
+(defun chomp (str)
```
Comment (bbatsov) [inline swift-mode.el]: You should not define unnamespaced functions. Btw, `subr-x.el` has exactly what you need.

File: swift-mode.el [hunk 2]
```diff
@@ -693,6 +693,115 @@ We try to constraint those lookups by reasonable number of lines.")
...
+  "Chomp leading and tailing whitespace from STR."
+  (replace-regexp-in-string (rx (or (: bos (* (any " \t\n")))
+                                    (: (* (any " \t\n")) eos)))
+                            ""
+                            str))
+
+(defun swift-mode-find-workspaces-for-directory(directory)
+  (setq command (concat "find " directory))
```
Comment (bbatsov) [inline swift-mode.el]: Working on the same variables from a few function is a very bad idea. Variables supposed to hold state between function invocations should be defined in advance and their semantics should be specified. Global state is evil!

File: swift-mode.el [hunk 3]
```diff
@@ -693,6 +693,115 @@ We try to constraint those lookups by reasonable number of lines.")
...
+  (setq command (concat command " -name \"*workspace\" -maxdepth 1"))
+  (setq workspaces (split-string
+                    (shell-command-to-string command)))
+  (mapcar 'chomp workspaces)
```
Comment (bbatsov) [inline swift-mode.el]: #'chomp is a better idea - would notify the compile this is a function

File: swift-mode.el [hunk 4]
```diff
@@ -693,6 +693,115 @@ We try to constraint those lookups by reasonable number of lines.")
...
+(defun swift-mode-find-projects-for-directory(directory)
+  (setq command (concat "find " directory))
+  (setq command (concat command " -name \"*xcodeproj\" -maxdepth 1"))
+  (setq projects (split-string
+                    (shell-command-to-string command)))
+  (mapcar 'chomp projects)
+  )
+
+(defun swift-mode-find-schemes-for-workspace(workspace)
+	(setq list (cdr (cdr (split-string
```
Comment (bbatsov) [inline swift-mode.el]: Incorrect indentation

File: swift-mode.el (hunk 1 above)
Comment (nicklanasa) [inline swift-mode.el]: @bbatsov Perfect, I'll make that change.

File: swift-mode.el (hunk 2 above)
Comment (nicklanasa) [inline swift-mode.el]: Specifically, which variables are you talking about. `directory`?

File: swift-mode.el (hunk 3 above)
Comment (nicklanasa) [inline swift-mode.el]: What exactly do you mean by this?

**Comment by nicklanasa** [comment]
Anybody else have any comments or concerns? Would love some more feedback.

**Comment by ap4y** [comment]
Hi @nicklanasa, I would recommend squashing commits into one to keep master clean. Also will be good if @bbatsov will review this PR one more time.

**Comment by nicklanasa** [comment]
@ap4y I can do that.

**Comment by nicklanasa** [comment]
@ap4y Rebased into 1 commit. Much cleaner.

**Comment by ap4y** [comment]
Thanks!

File: swift-mode.el [hunk 5]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+                        (concat " -workspace " workspace))))
+      (setq compilation-scroll-output t)
+      (pop-to-buffer "*compilation*"))))
+  
+(defun swift-mode-run-pod-install()
+  "Runs pod install."
+  (interactive)
+  (swift-mode-kill-compilation-if-exists)
+  (pop-to-buffer "*compilation*")
+  (setq compilation-scroll-output t)
```
Comment (bbatsov) [inline swift-mode.el]: You probably want just to let-bind those variables, as they are dynamic.

File: swift-mode.el [hunk 6]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+                                 (swift-mode-find-schemes-for-workspace workspace) nil t))
+                        (concat " -workspace " workspace))))
+      (setq compilation-scroll-output t)
+      (pop-to-buffer "*compilation*"))))
+  
+(defun swift-mode-run-pod-install()
+  "Runs pod install."
+  (interactive)
+  (swift-mode-kill-compilation-if-exists)
+  (pop-to-buffer "*compilation*")
```
Comment (bbatsov) [inline swift-mode.el]: Is this needed?

File: swift-mode.el [hunk 7]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+                          (concat " -workspace " workspace)))
+                 (concat " -destination 'id=" (concat (car (swift-mode-select-destination-id)) "'")))))
+      (swift-mode-kill-compilation-if-exists)
+      (setq compilation-scroll-output t)
+      (pop-to-buffer "*compilation*"))))
+
+(defun swift-mode-clean-workspace()
+  "Cleans the Xcode workspace using xcodebuild."
+  (interactive)
+  (progn
```
Comment (bbatsov) [inline swift-mode.el]: This seems redundant or the indentation is wrong.

File: swift-mode.el [hunk 8]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+                 (concat " -sdk " (swift-mode-select-sdk)))
+                (concat " -configuration " (swift-mode-select-build-config))))
+      (swift-mode-kill-compilation-if-exists)
+      (setq compilation-scroll-output t)
+      (pop-to-buffer "*compilation*"))))
+
+(defun swift-mode-test-workspace()
+  "Test the Xcode workspace using xcodebuild."
+  (interactive)
+  (progn
```
Comment (bbatsov) [inline swift-mode.el]: Same here.

File: swift-mode.el [hunk 9]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
                "warning: " (message) line-end))
      :modes swift-mode))
 
+;;; BUILDING
+
+(defun swift-mode-find-workspaces-for-directory(directory)
+  (mapcar 'string-trim (split-string (shell-command-to-string
```
Comment (bbatsov) [inline swift-mode.el]: #'string-trim

File: swift-mode.el [hunk 10]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+                                      (concat
```
Comment (bbatsov) [inline swift-mode.el]: There's also `format`.

File: swift-mode.el [hunk 11]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+
+(defun swift-mode-find-projects-for-directory(directory)
+  (mapcar 'string-trim (split-string (shell-command-to-string
+                                      (concat
+                                       (concat "find " (substring directory 0 -1))
+                                       (concat " -name \"*xcodeproj\" -maxdepth 1"))))))
+
+(defun swift-mode-find-schemes-for-workspace(workspace)
+  (mapcar 'string-trim
+          (cdr (cdr
```
Comment (bbatsov) [inline swift-mode.el]: Nested `cdr`s create pretty unreadable code.

File: swift-mode.el [hunk 12]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+                (split-string (shell-command-to-string
+                               (concat
+                                (concat "xcodebuild -workspace " workspace) " -list")) "\n")))))
+
+(defun swift-mode-kill-compilation-if-exists()
```
Comment (bbatsov) [inline swift-mode.el]: Not sure why this is needed.

File: swift-mode.el [hunk 13]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+(defvar swift-mode-platforms-list nil)
+
+(defun swift-mode-get-platform-list()
+  (setq swift-mode-platforms-list
+        (cdr (assoc 'devices
+                    (json-read-from-string
+                     (shell-command-to-string "xcrun simctl list devices -j"))))))
+
+(defun swift-mode-get-device-list(platform)
+  (let ((v (cdr (assoc-if #'(lambda(str) (string= platform str))
```
Comment (bbatsov) [inline swift-mode.el]: `v` is not a very informative variable name.

File: swift-mode.el [hunk 14]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+                          swift-mode-platforms-list))))
+    (mapcar (lambda (element)
+              (let ((list '()))
```
Comment (bbatsov) [inline swift-mode.el]: list is a bad var name as well :-)

File: swift-mode.el [hunk 15]
```diff
@@ -693,6 +693,175 @@ We try to constraint those lookups by reasonable number of lines.")
...
+                (let ((device (cdr element))
+                      (udid (cdr (car element))))
+                  (cons (concat (cdr (assoc 'name device)) (concat " - " udid)) list))))
+            (append v nil))))
+
+(defun swift-mode-select-destination-id()
+  (cdr (cdr (cdr
```
Comment (bbatsov) [inline swift-mode.el]: If you need just one elem - there's `nth`.

**Comment by nicklanasa** [comment]
Thanks for all the feedback, I'm still a noob with LISP but I'm learning so much! @bbatsov @ap4y

File: swift-mode.el (hunk 12 above)
Comment (nicklanasa) [inline swift-mode.el]: It seems to fix the never ending xcodebuild's. Maybe there is something else that could fix this but killing the `*compilation*` buffer fixes the issue.

**Comment by nicklanasa** [comment]
Added some cleanup and bug fixes. Should look better now.

**Pull request closed without merging**